### PR TITLE
Metrics for Dicom update

### PR DESCRIPTION
**Pull request by arunmk-ms**
## Description
Added metric for recording the number of instances updated

## Related issues
Addresses [102011](https://microsofthealth.visualstudio.com/Health/_boards/board/t/Medical%20Imaging/Stories/?workitem=102011)

## Testing
Manual testing and verification

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 1]
```diff
@@ -110,6 +110,7 @@ public partial class UpdateDurableFunction
                     Errors = input.Errors,
                     CreatedTime = input.CreatedTime ?? await context.GetCreatedTimeAsync(_options.RetryOptions),
                 });
+            _updateMeter.UpdatedInstances.Add(instanceWatermarks.Count);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:113]: this should be above line: 102

Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:113]: this should be above line: 102. Also could you add a unit test that we have called the meter

Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs]: Done

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/OpenTelemetryLabels.cs [hunk 2]
```diff
@@ -8,4 +8,5 @@ namespace Microsoft.Health.Dicom.Core.Features.Telemetry;
 public static class OpenTelemetryLabels
 {
     public const string BaseMeterName = "Microsoft.Health.Dicom";
+    public const string PlatformBaseMeterName = "Microsoft.Health.Platform.Dicom";
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Telemetry/OpenTelemetryLabels.cs:11]: Why a new base name?

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Diagnostics.Metrics;
+
+namespace Microsoft.Health.Dicom.Core.Features.Telemetry;
+public sealed class UpdateMeter : IDisposable
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs:10]: nit: missing newline between class and namespace

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+{
+    private readonly Meter _meter;
+
+    public UpdateMeter()
+    {
+        _meter = new Meter($"{OpenTelemetryLabels.PlatformBaseMeterName}.Update", "1.0");
+        UpdatedInstances = _meter.CreateCounter<int>(nameof(UpdatedInstances), description: "Count of instances updated successfully");
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs:17]: Should the metric just be called `Count` given the [OTel naming conventions](https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/#use-count-instead-of-pluralization)

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs:17]: Or if we anticipate we want to count other things like studies and series, then perhaps just `"instances"`?

File: src/Microsoft.Health.Dicom.Functions.UnitTests/Update/UpdateDurableFunctionTests.cs [hunk 5]
```diff
@@ -39,13 +46,25 @@ public UpdateDurableFunctionTests()
...
+            Options.Create(_jsonSerializerOptions),
+            _updateMeter);
+        InitializeMetricExporter();
+    }
+
+    private void InitializeMetricExporter()
+    {
+        _exportedItems = new List<Metric>();
+        _meterProvider = Sdk.CreateMeterProviderBuilder()
+            .AddMeter($"{OpenTelemetryLabels.PlatformBaseMeterName}.*")
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions.UnitTests/Update/UpdateDurableFunctionTests.cs:66]: nit: we're only interested in update metrics, so let's only add those meters

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 6]
```diff
@@ -99,6 +99,10 @@ public partial class UpdateDurableFunction
                     instanceWatermarks);
             }
 
+            if (instanceWatermarks.Count > 0)
+            {
+                _updateMeter.UpdatedInstances.Add(instanceWatermarks.Count);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:104]: So, where did we land on addressing possible duplicates?

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.cs [hunk 7]
```diff
@@ -33,7 +35,8 @@ public partial class UpdateDurableFunction
         IMetadataStore metadataStore,
         IFileStore fileStore,
         IUpdateInstanceService updateInstanceService,
-        IOptions<JsonSerializerOptions> jsonSerializerOptions)
+        IOptions<JsonSerializerOptions> jsonSerializerOptions,
+        UpdateMeter updateMeter)
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.cs:39]: nit: let's add the meter above the options

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs (hunk 6 above)
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:104]: Do we want to add any dimension to uniquely identify this entry?

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/OpenTelemetryLabels.cs (hunk 2 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Telemetry/OpenTelemetryLabels.cs]: Removed the base name as we discussed.

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs (hunk 4 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs:18]: Discussed offline.

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs (hunk 6 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:104]: We will be addressing the duplicates as part of a different work item.

Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs]: Added a dimension as you suggested.

**Review by wsugarman — CHANGES_REQUESTED** [review]
Let's hold off on this PR until we're clearer on the unique identifier

File: src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs [hunk 8]
```diff
@@ -99,6 +99,11 @@ public partial class UpdateDurableFunction
                     instanceWatermarks);
             }
 
+            if (instanceWatermarks.Count > 0)
+            {
+                _updateMeter.UpdatedInstances.Add(instanceWatermarks.Count,
+                    new KeyValuePair<string, object>("ExecutionId", context.NewGuid()));
+            }
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Orchestration.cs:106]: nit: newline

File: src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    public UpdateMeter()
+    {
+        _meter = new Meter($"{OpenTelemetryLabels.BaseMeterName}.Update", "1.0");
+        UpdatedInstances = _meter.CreateCounter<int>(nameof(UpdatedInstances), description: "Count of instances updated successfully");
+    }
+
+    public Counter<int> UpdatedInstances { get; }
+
+    public void Dispose() => _meter.Dispose();
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Telemetry/UpdateMeter.cs:23]: nit: put `=> meter.Dipose()` on new line like:
```csharp
public void Dispose()
    => _meter.Dispose();
```

**Review by wsugarman — APPROVED** [review]

**Review by wsugarman — APPROVED** [review]

**Review by bcarthic — APPROVED** [review]

**Pull request merged**